src/server/utilities.ts

```diff
@@ -232,6 +232,79 @@ namespace ts.server {
         }
     }
 
+    export function cleanExistingMap<T>(
+        existingMap: Map<T>,
+        onDeleteExistingValue: (key: string, existingValue: T) => void) {
+        if (existingMap) {
+            // Remove all
+            existingMap.forEach((existingValue, key) => {
+                existingMap.delete(key);
+                onDeleteExistingValue(key, existingValue);
+            });
+        }
+    }
+
+    export function mutateExistingMapWithNewSet<T>(
+        existingMap: Map<T>, newMap: Map<true>,
+        createNewValue: (key: string) => T,
+        onDeleteExistingValue: (key: string, existingValue: T) => void
+    ): Map<T> {
+        return mutateExistingMap(
+            existingMap, newMap,
+            // Same value if the value is set in the map
+            /*isSameValue*/(_existingValue, _valueInNewMap) => true,
+            /*createNewValue*/(key, _valueInNewMap) => createNewValue(key),
+            onDeleteExistingValue,
+            // Should never be called since we say yes to same values all the time
+            /*OnDeleteExistingMismatchValue*/(_key, _existingValue) => notImplemented()
+        );
+    }
+
+    export function mutateExistingMap<T, U>(
+        existingMap: Map<T>, newMap: Map<U>,
+        isSameValue: (existingValue: T, valueInNewMap: U) => boolean,
+        createNewValue: (key: string, valueInNewMap: U) => T,
+        onDeleteExistingValue: (key: string, existingValue: T) => void,
+        OnDeleteExistingMismatchValue: (key: string, existingValue: T) => void
+    ): Map<T> {
+        // If there are new values update them
+        if (newMap) {

```

Comment:
If this parameter truly is optional, its type should be `| undefined`.
Is it really important for performance that we use `undefined` to represent an empty map? It seems like if we always had the map defined, there wouldn't be any need to return a value from this function. Also, if the map were always mutated rather than replaced, it wouldn't be a big deal to allocate space for empty maps.
(Nit: Also, I would prefer an early return.)